pkgs/development/python-modules/keyutils/default.nix

```
@@ -16,7 +16,12 @@ buildPythonPackage rec {
     substituteInPlace setup.py --replace '"pytest-runner"' ""
   '';
 
+  postConfigure = ''
+    cython keyutils/_keyutils.pyx
+  '';
+
   buildInputs = [ keyutils ];
+  nativeBuildInputs = [ cython ];
   nativeCheckInputs = [ pytest pytest-runner ];
```

```suggestion
  nativeCheckInputs = [
    pytestCheckHook
  ];
```

Drop pytest-runner, it is already patched out in `postPatch` and I would be surprised if it were used during `checkPhase`.

Also use `pytestCheckHook` instead of relying on the implicit call to `python setup.py test`.

---

Running with this gives a circular dependency error:

```
error: builder for '/nix/store/3hf58mxgpjrcbmkps8b4v57kqlgdv6mx-python3.11-keyutils-0.6.drv' failed with exit code 2;
       last 10 log lines:
       > test/keyutils_test.py:23: in <module>
       >     import keyutils
       > keyutils/__init__.py:21: in <module>
       >     from . import _keyutils
       > E   ImportError: cannot import name '_keyutils' from partially initialized module 'keyutils' (most likely due to a circular import
) (/build/source/keyutils/__init__.py)
       > =========================== short test summary info ============================
       > ERROR test/keyutils_test.py
       > !!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
       > =============================== 1 error in 0.03s ===============================
       > /nix/store/j4w7yj8c4z7zg5y6bhhypnl456dwmrc5-stdenv-linux/setup: line 1596: pop_var_context: head of shell_variables not a function
 context
       For full logs, run 'nix log /nix/store/3hf58mxgpjrcbmkps8b4v57kqlgdv6mx-python3.11-keyutils-0.6.drv'.
```

Reverting to `nativeCheckInputs = [ pytest pytest-runner ]` seems to pass everything:

```
python3.11-keyutils> Executing setuptoolsCheckPhase
python3.11-keyutils> running pytest
python3.11-keyutils> /nix/store/2d12djgkcij9pgdzsm5214cj70j7f9wq-python3.11-setuptools-68.2.2/lib/python3.11/site-packages/setuptools/comma
nd/test.py:193: _DeprecatedInstaller: setuptools.installer and fetch_build_eggs are deprecated.
python3.11-keyutils> !!
python3.11-keyutils>         ********************************************************************************
python3.11-keyutils>         Requirements should be satisfied by a PEP 517 installer.
python3.11-keyutils>         If you are using pip, you can try `pip install --use-pep517`.
python3.11-keyutils>         ********************************************************************************
python3.11-keyutils> !!
python3.11-keyutils>   ir_d = dist.fetch_build_eggs(dist.install_requires)
python3.11-keyutils> /nix/store/2d12djgkcij9pgdzsm5214cj70j7f9wq-python3.11-setuptools-68.2.2/lib/python3.11/site-packages/setuptools/comma
nd/test.py:194: _DeprecatedInstaller: setuptools.installer and fetch_build_eggs are deprecated.
python3.11-keyutils> !!
python3.11-keyutils>         ********************************************************************************
python3.11-keyutils>         Requirements should be satisfied by a PEP 517 installer.
python3.11-keyutils>         If you are using pip, you can try `pip install --use-pep517`.
python3.11-keyutils>         ********************************************************************************
python3.11-keyutils> !!
python3.11-keyutils>   tr_d = dist.fetch_build_eggs(dist.tests_require or [])
python3.11-keyutils> /nix/store/2d12djgkcij9pgdzsm5214cj70j7f9wq-python3.11-setuptools-68.2.2/lib/python3.11/site-packages/setuptools/comma
nd/test.py:195: _DeprecatedInstaller: setuptools.installer and fetch_build_eggs are deprecated.
python3.11-keyutils> !!
python3.11-keyutils>         ********************************************************************************
python3.11-keyutils>         Requirements should be satisfied by a PEP 517 installer.
python3.11-keyutils>         If you are using pip, you can try `pip install --use-pep517`.
python3.11-keyutils>         ********************************************************************************
python3.11-keyutils> !!
python3.11-keyutils>   er_d = dist.fetch_build_eggs(
python3.11-keyutils> running egg_info
python3.11-keyutils> writing keyutils.egg-info/PKG-INFO
python3.11-keyutils> writing dependency_links to keyutils.egg-info/dependency_links.txt
python3.11-keyutils> writing top-level names to keyutils.egg-info/top_level.txt
python3.11-keyutils> reading manifest file 'keyutils.egg-info/SOURCES.txt'
python3.11-keyutils> reading manifest template 'MANIFEST.in'
python3.11-keyutils> adding license file 'LICENSE'
python3.11-keyutils> adding license file 'AUTHORS'
python3.11-keyutils> writing manifest file 'keyutils.egg-info/SOURCES.txt'
python3.11-keyutils> running build_ext
python3.11-keyutils> skipping 'keyutils/_keyutils.c' Cython extension (up-to-date)
python3.11-keyutils> copying build/lib.linux-aarch64-cpython-311/keyutils/_keyutils.cpython-311-aarch64-linux-gnu.so -> keyutils
python3.11-keyutils> ============================= test session starts ==============================
python3.11-keyutils> platform linux -- Python 3.11.5, pytest-7.4.2, pluggy-1.2.0
python3.11-keyutils> rootdir: /build/source
python3.11-keyutils> configfile: setup.cfg
python3.11-keyutils> collected 9 items                                                              
python3.11-keyutils> test/keyutils_test.py .........                                          [100%]
python3.11-keyutils> ============================== 9 passed in 1.52s ===============================
python3.11-keyutils> Finished executing setuptoolsCheckPhase
```

I'm not super versed in the Python ecosystem (and the Nix build system in this respect either) so I'm not too sure if this is working as originally intended or if there's something that could be fixed here. Thoughts?

---

The package has a native module that it tries to import. While running the tests it currently looks at the source package, not the built one.

The solution is to delete the source package before running the tests.

```
  preCheck = ''
    rm -rf keyutils
  '';
```

---

Thanks @mweinelt! That did it. All updated now and re-requesting.